sgl-kernel/csrc/cpu/gemm.cpp

```diff
@@ -378,6 +398,137 @@ void weight_packed_linear_kernel_impl(
+
+  const fVec one = fVec(1.f);
+  fVec X = fVec(input[0] + bias[0]);
+  X = one / (one + X.neg().exp_u20());
+  constexpr int kVecSize = bVec::size();
+  for (int d = 0; d < SIZE; d += kVecSize) {
+    bVec m_bvec = bVec::loadu(mul + d);
+    fVec m_fvec0, m_fvec1;
+    std::tie(m_fvec0, m_fvec1) = at::vec::convert_to_float(m_bvec);
+    m_fvec0 = m_fvec0*X;
+    m_fvec1 = m_fvec1*X;
+
+    bVec out_vec = convert_from_float_ext<scalar_t>(m_fvec0, m_fvec1);
+    out_vec.store(out + d);
+  }
+}
+
+template <typename scalar_t, bool has_bias>
+void tinygemm_kernel_f32(
+    const float* __restrict__ A,
+    const float* __restrict__ B,
+    scalar_t* __restrict__ C,
+    float* __restrict__ Ctmp,
+    const float* __restrict__ bias,
+    const scalar_t* __restrict__ post_mul,
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t lda,
+    int64_t ldb,
+    int64_t ldc,
+    bool brg) {
+
+    at::native::cpublas::brgemm(M, N, K, lda, ldb, ldc, /* add_C */ false, A, B, Ctmp);
+    // copy from Ctmp to C
+    for (int64_t m = 0; m < M; ++m) {
+      if constexpr (has_bias) {
+        if (post_mul != nullptr){
+            bias_sigmoid_mul<scalar_t>(C + m * K, Ctmp + m * N, bias, post_mul+m * K, K);
+        }else{
+            copy_add_stub(C + m * ldc, Ctmp + m * N, bias, N);
+        }
+      } else {
+        if (post_mul != nullptr){
+            sigmoid_mul<scalar_t>(C + m * K, Ctmp + m * N, post_mul+m * K, K);
+        }else{
+            copy_stub(C + m * ldc, Ctmp + m * N, N);
+        }
+      }
+    }
+    return;
+
+}
+
+template <typename scalar_t>
+void fma_linear_kernel_impl(
+    scalar_t* __restrict__ out,
+    float* __restrict__ mat1_f32,
+    float* __restrict__ mat2_f32,
+    const float* __restrict__ bias,
+    const scalar_t* __restrict__ post_mul,
+    int64_t M,
+    int64_t N,
+    int64_t K) {
+  constexpr int64_t BLOCK_M = block_size_m();
+  const int64_t MB = div_up(M, BLOCK_M);
+  // parallel on [MB, NB]
+
+  AT_DISPATCH_BOOL(bias != nullptr, has_bias, [&] {
+    at::parallel_for(0, MB, 0, [&](int64_t begin, int64_t end) {
+      int64_t mb{0};
+      data_index_init(begin, mb, MB);
+      // for brgemm, use float32 for accumulate
+      alignas(64) float Ctmp[BLOCK_M * N];
+
+      for (int64_t i = begin; i < end; ++i) {
+        UNUSED(i);
+        int64_t mb_start = mb * BLOCK_M;
+        int64_t mb_size = std::min(M - mb_start, BLOCK_M);
+        // bf16 -> fp32
+        tinygemm_kernel_f32<scalar_t, has_bias>(
+            /*   A */ mat1_f32 + mb_start * K,
+            /*   B */ mat2_f32,
+            /*   C */ post_mul==nullptr? out + mb_start * N: out + mb_start * K,
+            /* Ctmp*/ Ctmp,
+            /* bias*/ bias,
+            post_mul==nullptr? post_mul : post_mul + mb_start * K,
+            /*   M */ mb_size,
+            /*   N */ N,
+            /*   K */ K,
+            /* lda */ K,
+            /* ldb */ N,
+            /* ldc */ N,
+            /* brg */ true);
```

Comment:
separate this function into two parts and it will align better with existing code:

* first part is gemm NN for fp32
template <bool has_bias, int BLOCK_M, int BLOCK_N>
struct tinygemm_kernel_nn<float, has_bias, BLOCK_M, BLOCK_N> {};

* second part handled post ops.